File: android/app/src/main/java/com/brekeke/phone/utils/Ringtone.java
package com.brekeke.phone.utils;

import android.content.Context;
import android.media.AudioAttributes;
import android.media.AudioDeviceInfo;
import android.media.AudioManager;
import android.media.AudioManager.OnCommunicationDeviceChangedListener;
import android.media.AudioManager.OnModeChangedListener;
import android.media.MediaPlayer;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.text.TextUtils;
import android.util.Pair;
import com.facebook.react.bridge.NativeArray;
import com.facebook.react.bridge.WritableNativeArray;
import java.io.File;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

// utils to handle ringtone
// see the related part in rn js for reference

public class Ringtone {
  // ==========================================================================
  // init

  public static void init() {
    if (am != null) {
      return;
    }
    var ctx = Ctx.app();
    am = (AudioManager) ctx.getSystemService(Context.AUDIO_SERVICE);
    debug();
  }

  private static void debug() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
      return;
    }
    var l1 =
        new OnModeChangedListener() {
          @Override
          public void onModeChanged(int m) {
            var k1 = "onModeChanged:mode::";
            var k2 = k1 + "AudioManager.";
            switch (m) {
              case AudioManager.MODE_NORMAL:
                Emitter.debug(k2 + "MODE_NORMAL");
                break;
              case AudioManager.MODE_INVALID:
                Emitter.debug(k2 + "MODE_INVALID");
                break;
              case AudioManager.MODE_CURRENT:
                Emitter.debug(k2 + "MODE_CURRENT");
                break;
              case AudioManager.MODE_RINGTONE:
                Emitter.debug(k2 + "MODE_RINGTONE");
                break;
              case AudioManager.MODE_IN_CALL:
                Emitter.debug(k2 + "MODE_IN_CALL");
                break;
              case AudioManager.MODE_IN_COMMUNICATION:
                Emitter.debug(k2 + "MODE_IN_COMMUNICATION");
                break;
              case AudioManager.MODE_CALL_SCREENING:
                Emitter.debug(k2 + "MODE_CALL_SCREENING");
                break;
              default:
                Emitter.debug(k1 + m);
                break;
            }
          }
        };
    var l2 =
        new OnCommunicationDeviceChangedListener() {
          @Override
          public void onCommunicationDeviceChanged(AudioDeviceInfo device) {
            Emitter.debug(
                "onCommunicationDeviceChanged:AudioDeviceInfo::"
                    + device.getType()
                    + "::"
                    + device.getProductName());
          }
        };
    var ctx = Ctx.app();
    var e = ctx.getMainExecutor();
    am.addOnModeChangedListener(e, l1);
    am.addOnCommunicationDeviceChangedListener(e, l2);
  }

  // ==========================================================================
  // get all ringtone options

  public static NativeArray options() {
    var arr = new WritableNativeArray();
    for (var r : _static) {
      arr.pushString(r);
    }
    var pp = _picker();
    while (pp.hasNext()) {
      arr.pushString(pp.next());
    }
    try (var s = system()) {
      s.forEach(p -> arr.pushString(p.first));
    }
    return arr;
  }

  private static Stream<Pair<String, String>> system() {
    var ctx = Ctx.app();
    var rm = new RingtoneManager(ctx);
    rm.setType(RingtoneManager.TYPE_RINGTONE);
    var c = rm.getCursor();
    var iterator =
        new Iterator<Pair<String, String>>() {
          @Override
          public boolean hasNext() {
            return c.moveToNext();
          }

          @Override
          public Pair<String, String> next() {
            var title = c.getString(RingtoneManager.TITLE_COLUMN_INDEX);
            var uri = rm.getRingtoneUri(c.getPosition()).toString();
            return new Pair<>(title, uri);
          }
        };
    var spliterator =
        (Spliterator<Pair<String, String>>) Spliterators.spliteratorUnknownSize(iterator, 0);
    return StreamSupport.stream(spliterator, false).onClose(c::close);
  }

  // ==========================================================================
  // validate

  private static String[] _static = {"incallmanager_ringtone"};
  public static String _default = _static[0];
  private static String defaultFormat = ".mp3";

  // map ringtone url -> 1 (true) / 0 (false)
  private static Map<String, String> errors = new HashMap<>();

  private static String validate(String r) {
    if (TextUtils.isEmpty(r)) {
      return null;
    }
    if (_static(r)) {
      return r;
    }
    if (https(r)) {
      return r;
    }
    var pp = pickerPath(r);
    if (!pp.isEmpty()) {
      return pp;
    }
    try (var s = system()) {
      return s.filter(p -> p.first.equals(r)).map(p -> p.second).findFirst().orElse(null);
    }
  }

  private static String validateWithError(String r) {
    r = validate(r);
    if (TextUtils.isEmpty(r)) {
      return null;
    }
    if ("1".equals(errors.get(r))) {
      return null;
    }
    return r;
  }

  private static boolean _static(String r) {
    return Arrays.asList(_static).contains(r);
  }

  private static boolean https(String r) {
    return r.startsWith("https://");
  }

  private static String pickerPath(String filename) {
    var ctx = Ctx.app();
    // use the `Ringtones` folder name to sync with the @react-native-documents/picker library patch
    File file = new File(ctx.getFilesDir(), "Ringtones");
    File des = new File(file, filename + defaultFormat);
    if (!des.exists()) {
      return "";
    }
    return des.getAbsolutePath();
  }

  // get from push notification and validate
  public static String get(String r, String u, String t, String h, String p) {
    try {
      var v = validateWithError(r);
      if (!TextUtils.isEmpty(v)) {
        return v;
      }
      return get(u, t, h, p);
    } catch (Exception e) {
    }
    return _default;
  }

  // get from account and validate
  private static String get(String u, String t, String h, String p) {
    try {
      var a = Account.find(u, t, h, p);
      var r = validateWithError(a.getString("ringtone"));
      if (!TextUtils.isEmpty(r)) {
        return r;
      }
      r = validateWithError(a.getString("pbxRingtone"));
      if (!TextUtils.isEmpty(r)) {
        return r;
      }
    } catch (Exception e) {
    }
    return _default;
  }

  // ==========================================================================
  // play

  // static globally
  private static AudioManager am;

  // static but reinit each play
  private static Vibrator vib;
  private static MediaPlayer mp;

  private static int PLAY_TIMEOUT = 1000;
  private static Runnable onError;
  private static Handler h;

  private static Data d = new Data();

  public static boolean play(String r, String u, String t, String h, String p) {
    if (mp != null) {
      // return false if already playing
      return false;
    }
    d.set(r, u, t, h, p);
    int m = am.getRingerMode();
    if (m == AudioManager.RINGER_MODE_SILENT) {
      return true;
    }
    var ctx = Ctx.app();
    if (vib == null) {
      vib = (Vibrator) ctx.getSystemService(Context.VIBRATOR_SERVICE);
    }
    var pattern = new long[] {0, 1000, 1000};
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      vib.vibrate(VibrationEffect.createWaveform(pattern, new int[] {0, 255, 0}, 0));
    } else {
      vib.vibrate(pattern, 0);
    }
    if (m == AudioManager.RINGER_MODE_VIBRATE) {
      return true;
    }
    am.setMode(AudioManager.MODE_RINGTONE);
    playMp();
    return true;
  }

  private static void playMp() {
    try {
      playMpWithoutCatch(get(d.r, d.u, d.t, d.h, d.p));
    } catch (Exception e) {
      try {
        playMpWithoutCatch(get(d.u, d.t, d.h, d.p));
      } catch (Exception e2) {
        try {
          playMpWithoutCatch(_default);
        } catch (Exception e3) {
          Emitter.error("Ringtone playMp 3", e3.getMessage());
        }
        Emitter.error("Ringtone playMp 2", e2.getMessage());
      }
      Emitter.error("Ringtone playMp 1", e.getMessage());
    }
  }

  private static void playMpWithoutCatch(String r) throws Exception {
    stopMp();
    var ctx = Ctx.app();
    var attr =
        new AudioAttributes.Builder()
            .setContentType(AudioAttributes.CONTENT_TYPE_UNKNOWN)
            .setLegacyStreamType(AudioManager.STREAM_RING)
            .setUsage(AudioAttributes.USAGE_NOTIFICATION_RINGTONE)
            .build();
    // static
    if (_static(r)) {
      var res = ctx.getResources();
      var pkg = ctx.getPackageName();
      var id = res.getIdentifier(r, "raw", pkg);
      mp = MediaPlayer.create(ctx, id, attr, am.generateAudioSessionId());
      mp.setVolume(1.0f, 1.0f);
      mp.setLooping(true);
      mp.start();
      return;
    }
    // uri
    mp = new MediaPlayer();
    mp.setAudioAttributes(attr);
    mp.setDataSource(ctx, Uri.parse(r));
    mp.setVolume(1.0f, 1.0f);
    mp.setLooping(true);
    if (!https(r)) {
      mp.prepare();
      mp.start();
      return;
    }
    // https uri
    onError =
        () -> {
          if (!errors.containsKey(r)) {
            errors.put(r, "1");
          }
          playMp();
          stopOnError();
        };
    mp.setOnPreparedListener(
        m -> {
          m.start();
          errors.put(r, "0");
          stopOnError();
        });
    mp.setOnErrorListener(
        (v, w, s) -> {
          if (onError != null) {
            onError.run();
          }
          return true;
        });
    mp.prepareAsync();
    h = Ctx.h();
    h.postDelayed(onError, PLAY_TIMEOUT);
  }

  public static void stop() {
    d = new Data();
    stopVib();
    stopMp();
    stopOnError();
  }

  private static void stopVib() {
    if (vib != null) {
      try {
        vib.cancel();
      } catch (Exception e) {
      }
      vib = null;
    }
  }

  private static void stopMp() {
    if (mp != null) {
      try {
        mp.stop();
        mp.release();
      } catch (Exception e) {
      }
      mp = null;
    }
  }

  private static void stopOnError() {
    if (onError != null) {
      try {
        h.removeCallbacks(onError);
      } catch (Exception e) {
      }
      onError = null;
      h = null;
    }
  }

  // ==========================================================================
  // rn helpers

  public static void setAudioMode(int m) {
    switch (m) {
      case AudioManager.MODE_NORMAL:
        am.setMode(AudioManager.MODE_NORMAL);
        break;
      case AudioManager.MODE_RINGTONE:
        am.setMode(AudioManager.MODE_RINGTONE);
        break;
      case AudioManager.MODE_IN_CALL:
        am.setMode(AudioManager.MODE_IN_CALL);
        break;
      case AudioManager.MODE_IN_COMMUNICATION:
        am.setMode(AudioManager.MODE_IN_COMMUNICATION);
        break;
      case AudioManager.MODE_CALL_SCREENING:
        am.setMode(AudioManager.MODE_CALL_SCREENING);
        break;
      default:
        am.setMode(AudioManager.MODE_NORMAL);
        break;
    }
  }

  public static int getRingerMode() {
    return am.getRingerMode();
  }

  // ==========================================================================
  // picker
  private static Iterator<String> _picker() {
    try {
      var p = Storage.picker();
      return p.keys();
    } catch (Exception e) {
    }
    return Collections.emptyIterator();
  }
}

class Data {
  public String r;
  public String u;
  public String t;
  public String h;
  public String p;

  public void set(String r, String u, String t, String h, String p) {
    this.r = r;
    this.u = u;
    this.t = t;
    this.h = h;
    this.p = p;
  }
}

File: brekekephone/app/android/app/src/main/java/com/brekeke/phonedev/utils/Ringtone.kt
package com.brekeke.phonedev.utils

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.VibrationEffect
import android.os.Vibrator
import android.text.TextUtils
import android.util.Log
import android.util.Pair
import com.brekeke.phonedev.BrekekeUtils
import com.facebook.react.bridge.NativeArray
import com.facebook.react.bridge.WritableNativeArray
import java.io.File
import java.util.Arrays
import java.util.Collections
import java.util.Spliterators
import java.util.stream.Stream
import java.util.stream.StreamSupport

// utils to handle ringtone
// see the related part in rn js for reference

object Ringtone {
  val TAG = "[Ringtone]"

  @Volatile var shouldSkipPlayRingtone = false

  private var am: AudioManager? = null

  fun init() {
    if (am != null) return
    val ctx = Ctx.app()!!
    am = ctx.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    debug()
  }

  private fun debug() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return
    val l1 = AudioManager.OnModeChangedListener { m ->
      val k1 = "onModeChanged:mode::"
      val k2 = "${k1}AudioManager."
      when (m) {
        AudioManager.MODE_NORMAL -> Emitter.debug("${k2}MODE_NORMAL")
        AudioManager.MODE_INVALID -> Emitter.debug("${k2}MODE_INVALID")
        AudioManager.MODE_CURRENT -> Emitter.debug("${k2}MODE_CURRENT")
        AudioManager.MODE_RINGTONE -> Emitter.debug("${k2}MODE_RINGTONE")
        AudioManager.MODE_IN_CALL -> Emitter.debug("${k2}MODE_IN_CALL")
        AudioManager.MODE_IN_COMMUNICATION -> Emitter.debug("${k2}MODE_IN_COMMUNICATION")
        AudioManager.MODE_CALL_SCREENING -> Emitter.debug("${k2}MODE_CALL_SCREENING")
        else -> Emitter.debug(k1 + m)
      }
    }
    val l2 = AudioManager.OnCommunicationDeviceChangedListener { device: AudioDeviceInfo? ->
      if (device == null) {
        Emitter.debug("onCommunicationDeviceChanged:AudioDeviceInfo::null")
        return@OnCommunicationDeviceChangedListener
      }
      Emitter.debug(
          "onCommunicationDeviceChanged:AudioDeviceInfo::${device.type}::${device.productName}"
      )
    }
    val ctx = Ctx.app()!!
    val e = ctx.mainExecutor
    am!!.addOnModeChangedListener(e, l1)
    am!!.addOnCommunicationDeviceChangedListener(e, l2)
  }

  fun options(): NativeArray {
    val arr = WritableNativeArray()
    for (r in _static) arr.pushString(r)
    val pp = _picker()
    while (pp.hasNext()) arr.pushString(pp.next())
    system().use { s -> s.forEach { p -> arr.pushString(p.first) } }
    return arr
  }

  private fun system(): Stream<Pair<String, String>> {
    val ctx = Ctx.app()!!
    val rm = RingtoneManager(ctx)
    rm.setType(RingtoneManager.TYPE_RINGTONE)
    val c = rm.cursor
    val iterator =
        object : Iterator<Pair<String, String>> {
          override fun hasNext() = c.moveToNext()

          override fun next(): Pair<String, String> {
            val title = c.getString(RingtoneManager.TITLE_COLUMN_INDEX)
            val uri = rm.getRingtoneUri(c.position).toString()
            return Pair(title, uri)
          }
        }
    val spliterator = Spliterators.spliteratorUnknownSize(iterator, 0)
    return StreamSupport.stream(spliterator, false).onClose { c.close() }
  }

  private val _static = arrayOf("incallmanager_ringtone")
  val _default = _static[0]
  private val defaultFormat = ".mp3"
  private val errors: MutableMap<String, String> = HashMap()

  private fun validate(r: String?): String? {
    if (TextUtils.isEmpty(r)) return null
    if (_systemDefault(r!!)) {
      val u = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_RINGTONE)
      return u.toString()
    }
    if (_static(r)) return r
    if (https(r)) return r
    val pp = pickerPath(r)
    if (pp.isNotEmpty()) return pp
    system().use { s ->
      return s.filter { p -> p.first == r }.map { p -> p.second }.findFirst().orElse(null)
    }
  }

  private fun validateWithError(r: String?): String? {
    val v = validate(r) ?: return null
    if ("1" == errors[v]) return null
    return v
  }

  private fun _static(r: String) = Arrays.asList(*_static).contains(r)

  private fun _systemDefault(r: String) = "--" == r

  private fun https(r: String) = r.startsWith("https://")

  private fun pickerPath(filename: String): String {
    val ctx = Ctx.app()!!
    val file = File(ctx.filesDir, "Ringtones")
    val des = File(file, filename + defaultFormat)
    if (!des.exists()) return ""
    return des.absolutePath
  }

  fun get(r: String?, u: String?, t: String?, h: String?, p: String?): String {
    return try {
      val v = validateWithError(r)
      if (!TextUtils.isEmpty(v)) return v!!
      get(u, t, h, p)
    } catch (_: Exception) {
      _default
    }
  }

  private fun get(u: String?, t: String?, h: String?, p: String?): String {
    return try {
      val a = Account.find(u, t, h, p)!!
      var r = validateWithError(a.getString("ringtone"))
      if (!TextUtils.isEmpty(r)) return r!!
      r = validateWithError(a.getString("pbxRingtone"))
      if (!TextUtils.isEmpty(r)) return r!!
      _default
    } catch (_: Exception) {
      _default
    }
  }

  private var vib: Vibrator? = null
  private var mp: MediaPlayer? = null
  private const val PLAY_TIMEOUT = 1000L
  private var onError: Runnable? = null
  private var handler: Handler? = null
  private var d = Data()

  fun play(r: String?, u: String?, t: String?, h: String?, p: String?): Boolean {
    if (BrekekeUtils.anyCallAnswered()) {
      Emitter.debug("[Ringtone] Skip playing: another call is already answered.")
      return false
    }
    Log.d(TAG, "play: shouldSkipPlayRingtone = $shouldSkipPlayRingtone")
    if (shouldSkipPlayRingtone) {
      Emitter.debug("[Ringtone] Skip playing: a call is already connected.")
      return false
    }
    if (mp != null) {
      Emitter.debug("[Ringtone] Ringtone is playing")
      return false
    }
    d.set(r, u, t, h, p)
    val m = am!!.ringerMode
    if (m == AudioManager.RINGER_MODE_SILENT) return true
    val ctx = Ctx.app()!!
    if (vib == null) vib = ctx.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
    val pattern = longArrayOf(0, 1000, 1000)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      vib!!.vibrate(VibrationEffect.createWaveform(pattern, intArrayOf(0, 255, 0), 0))
    } else {
      @Suppress("DEPRECATION") vib!!.vibrate(pattern, 0)
    }
    if (m == AudioManager.RINGER_MODE_VIBRATE) return true
    am!!.mode = AudioManager.MODE_RINGTONE
    playMp()
    return true
  }

  private fun playMp() {
    try {
      playMpWithoutCatch(get(d.r, d.u, d.t, d.h, d.p))
    } catch (e: Exception) {
      try {
        playMpWithoutCatch(get(d.u, d.t, d.h, d.p))
      } catch (e2: Exception) {
        try {
          playMpWithoutCatch(_default)
        } catch (e3: Exception) {
          Emitter.error("Ringtone playMp 3", e3.message)
        }
        Emitter.error("Ringtone playMp 2", e2.message)
      }
      Emitter.error("Ringtone playMp 1", e.message)
    }
  }

  @Throws(Exception::class)
  private fun playMpWithoutCatch(r: String) {
    stopMp()
    val ctx = Ctx.app()!!
    val attr =
        AudioAttributes.Builder()
            .setContentType(AudioAttributes.CONTENT_TYPE_UNKNOWN)
            .setLegacyStreamType(AudioManager.STREAM_RING)
            .setUsage(AudioAttributes.USAGE_NOTIFICATION_RINGTONE)
            .build()
    if (_static(r)) {
      val res = ctx.resources
      val pkg = ctx.packageName
      val id = res.getIdentifier(r, "raw", pkg)
      mp = MediaPlayer.create(ctx, id, attr, am!!.generateAudioSessionId())
      mp!!.setVolume(1.0f, 1.0f)
      mp!!.isLooping = true
      mp!!.start()
      return
    }
    mp = MediaPlayer()
    mp!!.setAudioAttributes(attr)
    mp!!.setDataSource(ctx, Uri.parse(r))
    mp!!.setVolume(1.0f, 1.0f)
    mp!!.isLooping = true
    if (!https(r)) {
      mp!!.prepare()
      mp!!.start()
      return
    }
    onError = Runnable {
      if (!errors.containsKey(r)) errors[r] = "1"
      playMp()
      stopOnError()
    }
    mp!!.setOnPreparedListener { m ->
      m.start()
      errors[r] = "0"
      stopOnError()
    }
    mp!!.setOnErrorListener { _, _, _ ->
      onError?.run()
      true
    }
    mp!!.prepareAsync()
    handler = Ctx.h()
    handler!!.postDelayed(onError!!, PLAY_TIMEOUT)
  }

  fun stop() {
    d = Data()
    stopVib()
    stopMp()
    stopOnError()
  }

  private fun stopVib() {
    vib?.let {
      try {
        it.cancel()
      } catch (_: Exception) {}
      vib = null
    }
  }

  private fun stopMp() {
    mp?.let {
      try {
        it.stop()
        it.release()
      } catch (_: Exception) {}
      mp = null
    }
  }

  private fun stopOnError() {
    onError?.let {
      try {
        handler!!.removeCallbacks(it)
      } catch (_: Exception) {}
      onError = null
      handler = null
    }
  }

  fun setAudioMode(m: Int) {
    when (m) {
      AudioManager.MODE_NORMAL -> am!!.mode = AudioManager.MODE_NORMAL
      AudioManager.MODE_RINGTONE -> am!!.mode = AudioManager.MODE_RINGTONE
      AudioManager.MODE_IN_CALL -> am!!.mode = AudioManager.MODE_IN_CALL
      AudioManager.MODE_IN_COMMUNICATION -> am!!.mode = AudioManager.MODE_IN_COMMUNICATION
      AudioManager.MODE_CALL_SCREENING -> am!!.mode = AudioManager.MODE_CALL_SCREENING
      else -> am!!.mode = AudioManager.MODE_NORMAL
    }
  }

  fun getRingerMode(): Int = am!!.ringerMode

  private fun _picker(): Iterator<String> {
    return try {
      val p = Storage.picker()
      p.keys()
    } catch (_: Exception) {
      Collections.emptyIterator()
    }
  }
}

class Data {
  var r: String? = null
  var u: String? = null
  var t: String? = null
  var h: String? = null
  var p: String? = null

  fun set(r: String?, u: String?, t: String?, h: String?, p: String?) {
    this.r = r
    this.u = u
    this.t = t
    this.h = h
    this.p = p
  }
}
